File: day-14/tree.cpp
#include <bits/stdc++.h>
using namespace std;

class Node{
  int diameterHelper(Node* root){
    if(root == nullptr){
      return 0;
    }
    return max(diameterHelper(root->left), diameterHelper(root->right))+1;
  }
public:
  int data;
  Node* left, *right;
  Node* createNode(int val){
    Node* newNode = new Node;
    newNode->data = val;
    newNode->left = newNode->right = nullptr;
    return newNode;
  }

  void levelOreder(Node* root){
    queue<Node*> qu;
    qu.push(root);
    while(!qu.empty()){
      Node* temp = qu.front();
      cout << temp->data << " ";
      qu.pop();
      if(temp->left != nullptr){
        qu.push(temp->left);
      }
      if(temp->right != nullptr){
        qu.push(temp->right);
      }
    }
  }

  void reverseLevelOreder(Node* root){
    queue<Node*> qu;
    stack<Node*> st;
    qu.push(root);
    
    while(!qu.empty()){
      Node* temp = qu.front();
      st.push(temp);
      qu.pop();
      if(temp->right != nullptr){
        qu.push(temp->right);
      }
      if(temp->left != nullptr){
        qu.push(temp->left);
      }
    }
    while(!st.empty()){
      cout << st.top()->data << " ";
      st.pop();
    }
  }

  int diameter(Node* root){
    if(root == nullptr){
      return 0;
    }
    int left = diameterHelper(root->left);
    int right = diameterHelper(root->right);
    return left+right+1;
  }

  void postorder(Node* root){
    stack<Node*> st;
    Node* curr = root;
    while(curr || !st.empty()){
      if(curr){
        st.push(curr);
        curr = curr->left;
      } else{
        Node* top = st.top()->right;
        if(top == nullptr){
          top = st.top();
          st.pop();
          cout<<top->data << " ";
          while(!st.empty() && top == st.top()->right){
            top = st.top();
            st.pop();
            cout << top->data << " ";
          }
        } else{
          curr = top;
        }
      }
    }
  }
  
  void inorder(Node* root){
    if(root != nullptr){
      inorder(root->left);
      cout << root->data << " ";
      inorder(root->right);
    }
  }

};

int main(){

  Node* root = nullptr;
  root = root->createNode(10);
  root->right = root->createNode(20);
  root->left = root->createNode(5);
  root->right->left = root->createNode(15);
  root->right->right = root->createNode(30);
  root->right->right->right = root->createNode(40);
  cout << root->diameter(root);

  return 0;
}

File: day-58/Main/src/Tree.java
import java.util.*;

class Node {
    int data;
    Node left, right;

    Node(int val) {
        data = val;
        left = right = null;
    }
}

public class Tree {
    Node root;

    Tree() {
        root = null;
    }

    void inorder() {
        Stack<Node> s = new Stack<Node>();
        while (true) {
            while (root != null) {
                s.push(root);
                root = root.left;
            }
            if (s.isEmpty())
                break;
            root = s.pop();
            System.out.print(root.data + " ");
            root = root.right;
        }
    }

    void postorder() {
        Stack<Node> s = new Stack<Node>();
        Node prev = null;
        do {
            while (root != null) {
                s.push(root);
                root = root.left;
            }
            while (root == null && !s.isEmpty()) {
                root = s.peek();
                if (root.right == null || root.right == prev) {
                    System.out.print(root.data + " ");
                    prev = root;
                    s.pop();
                    root = null;
                } else {
                    root = root.right;
                }
            }
        } while (!s.isEmpty());
    }

    public static void main(String[] args) throws Exception {
        Tree t1 = new Tree();
        t1.root = new Node(1);
        t1.root.left = new Node(2);
        t1.root.left.left = new Node(4);
        t1.root.left.right = new Node(5);
        t1.root.right = new Node(3);
        t1.root.right.left = new Node(6);
        t1.root.right.right = new Node(7);
        t1.postorder();

    }
}
